### PR TITLE
DX: Update PHPStan

### PR DESCRIPTION
**Pull request by julienfalque**

### CODE REVIEW

File: src/Doctrine/Annotation/Tokens.php [hunk 1]
```diff
@@ -272,7 +272,7 @@ public function offsetUnset($index): void
         }
 
         $max = \count($this) - 1;
-        while ($index < $max) {
+        while ($index < $max && isset($this[$index + 1])) {
```
Comment (SpacePossum) [inline src/Doctrine/Annotation/Tokens.php:275]: I'm not a fan of this type of change., From the app view and domain we know for sure that `$this[$index + 1]` is always set, so the check is overcomplete. I wouldn't change valid code just because some SCA tooling triggers a fails issue.

File: src/Tokenizer/Token.php [hunk 2]
```diff
@@ -183,10 +183,7 @@ public function equals($other, bool $caseSensitive = true): bool
             }
         }
 
-        // detect unknown keys
-        unset($otherPrototype[0], $otherPrototype[1]);
-
-        return empty($otherPrototype);
+        return true;
```
Comment (SpacePossum) [inline src/Tokenizer/Token.php:186]: can you tell a bit about why this change?

File: src/Doctrine/Annotation/Tokens.php (hunk 1 above)
Comment (julienfalque) [reply src/Doctrine/Annotation/Tokens.php:275]: Investigating this a bit more, it seems `Tokens` and/or `SplFixedArray` have some weird behavior I don't fully understand. I tried the following:
```php
$tokens = Tokens::fromCode('<?php exit;');
unset($tokens[0]);
var_export($tokens);
```
This is what I get:
```php
PhpCsFixer\Tokenizer\Tokens::__set_state(array(
   'blockStartCache' => 
  array (
  ),
   'blockEndCache' => 
  array (
  ),
   'codeHash' => '2321664371',
   'changed' => true,
   'foundTokenKinds' => 
  array (
    390 => 0,
    323 => 1,
    ';' => 1,
  ),
   0 => NULL,
   1 => 
  PhpCsFixer\Tokenizer\Token::__set_state(array(
     'content' => 'exit',
     'id' => 323,
     'isArray' => true,
     'changed' => false,
  )),
   2 => 
  PhpCsFixer\Tokenizer\Token::__set_state(array(
     'content' => ';',
     'id' => NULL,
     'isArray' => false,
     'changed' => false,
  )),
))
```
This is not what I expected when looking at the implementation of `Tokens::offsetUnset()`.

Then I updated that method:
```php
    public function offsetUnset($index): void
    {
        throw new LogicException('Not implemented.');
    }
```
Same result and all tests seem to pass 🤨️

Anyway, `SplFixedArray` seems to always allow `null` as value somehow, so I suppose the error reported by PHPStan should be considered; but since I don't really understand why yet, I'd rather ignore it for now. What do you think?

File: src/Tokenizer/Token.php (hunk 2 above)
Comment (julienfalque) [reply src/Tokenizer/Token.php]: The `@param` annotation ensures that when `$other` is an array, it only has key `0` and optionally key `1`. This means that at this point, `$otherPrototype` cannot have other keys than `0` and `1` so unsetting them always results in an empty array, making the function always return `true`.

But this is only true when using static analysis and this method is not `@internal` so I reverted the change and ignored the error instead.

File: src/Doctrine/Annotation/Tokens.php (hunk 1 above)
Comment (SpacePossum) [reply src/Doctrine/Annotation/Tokens.php:275]: I think you confused the `PhpCsFixer\Tokenizer\Tokens` with `PhpCsFixer\Doctrine\Annotation\Tokens` in your example, which I totally get as I thought we were looking at the first as well ;)

Comment (SpacePossum) [reply src/Doctrine/Annotation/Tokens.php:275]: >tests/Doctrine/Annotation/TokensTest.php

Is lacking a lot of test on the functionality the collection delivers to be fair.
I've added this (rough and ugly test) locally to it to test the upper boundary `unset` and it seems fine:

```php
    public function testCreateAndUnset(): void
    {
        $docComment = '/** @var';

        $token = new Token([T_DOC_COMMENT, $docComment.'*/']);
        $tokens = Tokens::createFromDocComment($token);

        static::assertCount(4, $tokens);

        unset($tokens[3]);

        static::assertCount(3, $tokens);
        static::assertSame($docComment, $tokens->getCode());
    }
```

Comment (julienfalque) [reply src/Doctrine/Annotation/Tokens.php]: Well… that explains a lot 😂️

Change reverted, the error is now ignored.

File: src/Tokenizer/Token.php (hunk 2 above)
Comment (SpacePossum) [reply src/Tokenizer/Token.php]: check :+1:   thanks for the details :)

**Review by SpacePossum — APPROVED** [review]

File: src/Doctrine/Annotation/Tokens.php (hunk 1 above)
Comment (julienfalque) [reply src/Doctrine/Annotation/Tokens.php]: > > tests/Doctrine/Annotation/TokensTest.php
>
> Is lacking a lot of test on the functionality the collection delivers to be fair.

I noticed that as well. But since PHP now supports native attributes, I think Doctrine Annotations will become obsolete and I intend to deprecate all related fixers soon™, so maybe improving those tests is not worth the hassle. What do you think?

Comment (SpacePossum) [reply src/Doctrine/Annotation/Tokens.php]: Yeah I agree :+1: 
So extending the tests is no requirement in the PR what so ever. If someone would make a PR to add more coverage I would take a look for it, but if not.. well there are no open issues on it so all good :)
